### PR TITLE
Add bank compression executable

### PR DESCRIPTION
**Pull request by cdcapano**
This patch adds pycbc_bank_compress, which loads a template bank from an hdf file, compresses it, and saves back out to an hdf file. Two types of compression are possible: one based on the chirp mass estimate and one based on the t(f) estimated from the waveform using the SPA. This patch adds the SPA algorithm to compress.py, and makes the mchirp have more similar calling, return values. (Namely, both functions now return frequencies rather than indices.) In addition, the compress executable has an option for fixing the frequency points for all waveforms using the chirp estimate of the lowest mass waveform. There are thus 3 possible ways to compress the waveforms.

This patch also fixes a bug in FilterBank that caused it to ignore the dtype setting.

Speaking of dtypes, currently I'm forcing everything to be double precision, as the linear interpolation in decompress relies on using doubles. If the decompression can be made to handle single precision, then we could change this so that it can save single precision amp. & phase.

I've tested all 3 ways of compression on a subset of templates, and it seems to work (overlaps > 0.99 for all 3). I'm running a test on the full bank now; will update this with results.

### CODE REVIEW

**Comment by ahnitz** [comment]
@cdcapano See this pull request. The values are not currently being stored correctly. https://github.com/cdcapano/pycbc/pull/1

**Comment by cdcapano** [comment]
I ran a test on this yesterday trying both the mchirp and spa algorithms. We're not getting quite as good compression as we had hoped, though it's not terrible. Using the chirp mass compression, I get 127164516 total frequency points over the entire bank, so about 3GB total using double precision. The spa compression does a bit better, yielding 112002431 points, or about 2.7GB at double precision. However, the spa compression does pretty bad when decompressed. Here is a plot of mismatch between the decompressed spa waveforms and the full waveform vs chirp mass:

https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/waveform_compression_tests/spa-mm_mchirp_eta.png

The worst one has a mismtch of 20%. The chirp mass does much better, although there are still several points with > 1% mismatch:

https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/waveform_compression_tests/mchirp-mm_mchirp_eta.png

I took a look at the worst decompressed template by the spa compression. Top axis is real part of htilde vs. frequency; bottom is imaginary. Black is the full waveform, red is the decompressed:

https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/waveform_compression_tests/spa_decomp_compare-185761.png

In the bottom plot I've put x's where the compressed sample points are. The chirp mass compression for this particular template is much better (< 1% mismatch), but it looks to be over sapling it. Here is the same plot for the chirp mass decompression (ignore the green and cyan curves):

https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/waveform_compression_tests/mchirp_spa_compare-185761.png

I don't think the problem with the spa compression is due to  the number of points used, but where they are located. Since we are using linear interpolation, I think it might be necessary to ensure that the sample points occur at the peak in the amplitude. If you imagine shifting the sample points in the spa compression plot to the left, then it seems as though we should get a better result. I haven't checked yet, but I'd be willing to bet it's what's affecting the chirp mass compression for the templates with > 1% mismatch too. I'll try this tomorrow. Please hold off on merging this patch until then.

File: bin/pycbc_compress_bank [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    output['compressed_waveforms/frequency'] = fixed_points
+else:
+    sample_points = None
+    
+# get the compressed sample points for each template
+logging.info("getting compressed amplitude and phase")
+for ii in range(imin, imax):
+    htilde = bank[ii]
+    tmplt = bank.table[ii]
+    kmax = kmaxes[ii]
```
Comment (ahnitz) [inline bin/pycbc_compress_bank]: Any reason you aren't using the kmax that the template already calculates (and which might be different from SEOBNR RD?)

File: bin/pycbc_compress_bank [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                     "generating waveforms prior to compressing.")
+parser.add_argument("--tmplt-index", nargs=2, type=int, default=None,
+                help="Only generate compressed waveforms for the given "
+                     "indices in the bank file. Must provide both a start "
+                     "and a stop index. Default is to compress all of the "
+                     "templates in the bank file.")
+parser.add_argument("--compression-algorithm", type=str, default="mchirp",
+                help="The compression algorithm to use for selecting "
+                     "frequency points. Options are %s. Default is mchirp." %(
+                     ', '.join(compress.compression_algorithms.keys())))
```
Comment (titodalcanton) [inline bin/pycbc_compress_bank]: Hi Collin, you can use ArgumentParser's `choices` kwarg here, see other PyCBC scripts for examples.

**Comment by ahnitz** [comment]
Re-open when this gets updated.

**Pull request closed without merging**